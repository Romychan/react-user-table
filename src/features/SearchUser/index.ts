export { SearchUser } from './ui/SearchUser';
