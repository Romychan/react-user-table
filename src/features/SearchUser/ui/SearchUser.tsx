import { Icon } from '~/shared/ui/Icon';
import { InputGroup } from '~/shared/ui/InputGroup';
import { useDebounceCallback } from '~/shared/lib/hooks';

import styles from './SearchUser.module.scss';

interface ISearchUserProps {
  /**
   * A callback that will be called when the user enters a value to search for
   *
   * @param newValue New search value for filtering
   */
  onValueSearch: (newValue: string) => void;
}

/** This component is used to search for the user */
export const SearchUser = ({ onValueSearch }: ISearchUserProps) => {
  const handleSearchDebounce = useDebounceCallback<string>((newValue) =>
    onValueSearch(newValue),
  );

  return (
    <InputGroup
      id="search-user"
      name="search-user"
      onChange={(event) => handleSearchDebounce(event.target.value)}
      variant="filled"
      startContent={<Icon name="search" />}
      placeholder="Search by name..."
      className={styles.container}
      data-testid="search-user"
    />
  );
};
