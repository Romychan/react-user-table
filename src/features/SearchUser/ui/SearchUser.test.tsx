import {
  screen,
  render,
  RenderResult,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { vi } from 'vitest';

import { SearchUser } from './SearchUser';

describe('SearchUser', () => {
  let container: RenderResult;
  const handleChange = vi.fn();

  beforeEach(() => {
    container = render(<SearchUser onValueSearch={handleChange} />);
  });

  it('should render correctly', () => {
    const searchElement = screen.getByTestId('search-user');

    expect(searchElement).toBeInTheDocument();
  });

  it('should debounce value correctly', async () => {
    const searchElement = screen.getByTestId('search-user');

    fireEvent.change(searchElement, { target: { value: 'test' } });
    fireEvent.change(searchElement, { target: { value: 'test123' } });
    fireEvent.change(searchElement, { target: { value: '123' } });

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith('123');
    });
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
