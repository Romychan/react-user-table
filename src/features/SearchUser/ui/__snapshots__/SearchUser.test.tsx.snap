// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchUser > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_30c746 _container_c961cd"
      >
        <div
          class="_element_30c746 _startContent_30c746"
        >
          <svg
            class="icon icon-search"
            data-testid="icon"
            height="16"
            width="16"
          >
            <use
              xlink:href="/images/icons.svg#search"
            />
          </svg>
        </div>
        <input
          autocomplete="off"
          class="_input_95237c _filled_95237c"
          data-testid="search-user"
          id="search-user"
          name="search-user"
          placeholder="Search by name..."
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_30c746 _container_c961cd"
    >
      <div
        class="_element_30c746 _startContent_30c746"
      >
        <svg
          class="icon icon-search"
          data-testid="icon"
          height="16"
          width="16"
        >
          <use
            xlink:href="/images/icons.svg#search"
          />
        </svg>
      </div>
      <input
        autocomplete="off"
        class="_input_95237c _filled_95237c"
        data-testid="search-user"
        id="search-user"
        name="search-user"
        placeholder="Search by name..."
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
